File: apps/server/pkg/interfaces/http/handler/oauth_store.go
package handler

import (
	"context"
	"encoding/json"
	"log/slog"
	"net/http"
	"time"

	"github.com/lelemon/server/pkg/domain/entity"
	"github.com/lelemon/server/pkg/domain/repository"
)

// OAuthStoreHandler exposes repository.OAuthStore over a single internal RPC endpoint so the MCP
// authorization server (mcify, running out-of-process) can persist its clients/codes/tokens
// without direct DB access. Protected by ServiceAuth (shared secret). The JSON wire shapes match
// mcify's `OAuthStore` row types 1:1 (camelCase), so the TS adapter is a thin pass-through.
type OAuthStoreHandler struct {
	store repository.OAuthStore
}

func NewOAuthStoreHandler(store repository.OAuthStore) *OAuthStoreHandler {
	return &OAuthStoreHandler{store: store}
}

// ── Wire DTOs (== mcify OAuthStore rows) ─────────────────────────────────────────

type clientDTO struct {
	ClientID                string    `json:"clientId"`
	ClientName              *string   `json:"clientName"`
	RedirectURIs            []string  `json:"redirectUris"`
	GrantTypes              []string  `json:"grantTypes"`
	TokenEndpointAuthMethod string    `json:"tokenEndpointAuthMethod"`
	Scope                   *string   `json:"scope"`
	CreatedAt               time.Time `json:"createdAt"`
}

type authCodeDTO struct {
	CodeHash            string     `json:"codeHash"`
	ClientID            string     `json:"clientId"`
	SubjectKey          string     `json:"subjectKey"`
	RedirectURI         string     `json:"redirectUri"`
	CodeChallenge       string     `json:"codeChallenge"`
	CodeChallengeMethod string     `json:"codeChallengeMethod"`
	Scope               *string    `json:"scope"`
	ExpiresAt           time.Time  `json:"expiresAt"`
	ConsumedAt          *time.Time `json:"consumedAt,omitempty"`
}

type accessTokenDTO struct {
	TokenHash  string     `json:"tokenHash"`
	ClientID   string     `json:"clientId"`
	SubjectKey string     `json:"subjectKey"`
	Scope      *string    `json:"scope"`
	ExpiresAt  time.Time  `json:"expiresAt"`
	RevokedAt  *time.Time `json:"revokedAt,omitempty"`
}

type refreshTokenDTO struct {
	ID         string     `json:"id"`
	TokenHash  string     `json:"tokenHash"`
	ClientID   string     `json:"clientId"`
	SubjectKey string     `json:"subjectKey"`
	Scope      *string    `json:"scope"`
	ExpiresAt  time.Time  `json:"expiresAt"`
	ConsumedAt *time.Time `json:"consumedAt,omitempty"`
}

func clientToDTO(c *entity.OAuthClient) *clientDTO {
	if c == nil {
		return nil
	}
	return &clientDTO{
		ClientID: c.ClientID, ClientName: c.ClientName, RedirectURIs: c.RedirectURIs,
		GrantTypes: c.GrantTypes, TokenEndpointAuthMethod: c.TokenEndpointAuthMethod,
		Scope: c.Scope, CreatedAt: c.CreatedAt,
	}
}

func (d *clientDTO) toEntity() *entity.OAuthClient {
	return &entity.OAuthClient{
		ClientID: d.ClientID, ClientName: d.ClientName, RedirectURIs: d.RedirectURIs,
		GrantTypes: d.GrantTypes, TokenEndpointAuthMethod: d.TokenEndpointAuthMethod,
		Scope: d.Scope, CreatedAt: d.CreatedAt,
	}
}

func accessToDTO(t *entity.OAuthAccessToken) *accessTokenDTO {
	if t == nil {
		return nil
	}
	return &accessTokenDTO{
		TokenHash: t.TokenHash, ClientID: t.ClientID, SubjectKey: t.SubjectKey,
		Scope: t.Scope, ExpiresAt: t.ExpiresAt, RevokedAt: t.RevokedAt,
	}
}

func refreshToDTO(t *entity.OAuthRefreshToken) *refreshTokenDTO {
	if t == nil {
		return nil
	}
	return &refreshTokenDTO{
		ID: t.ID, TokenHash: t.TokenHash, ClientID: t.ClientID, SubjectKey: t.SubjectKey,
		Scope: t.Scope, ExpiresAt: t.ExpiresAt, ConsumedAt: t.ConsumedAt,
	}
}

// ── RPC envelope ─────────────────────────────────────────────────────────────────

type oauthRPCRequest struct {
	Op           string           `json:"op"`
	Client       *clientDTO       `json:"client"`
	Code         *authCodeDTO     `json:"code"`
	AccessToken  *accessTokenDTO  `json:"accessToken"`
	RefreshToken *refreshTokenDTO `json:"refreshToken"`
	ClientID     string           `json:"clientId"`
	ClientName   string           `json:"clientName"`
	CodeHash     string           `json:"codeHash"`
	TokenHash    string           `json:"tokenHash"`
	ID           string           `json:"id"`
	RotatedToID  string           `json:"rotatedToId"`
	SubjectKey   string           `json:"subjectKey"`
}

func (h *OAuthStoreHandler) Handle(w http.ResponseWriter, r *http.Request) {
	var req oauthRPCRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, `{"error":"invalid request"}`, http.StatusBadRequest)
		return
	}
	ctx := r.Context()

	result, err := h.dispatch(ctx, &req)
	if err != nil {
		slog.Error("oauth store rpc failed", "op", req.Op, "error", err)
		http.Error(w, `{"error":"store operation failed"}`, http.StatusInternalServerError)
		return
	}
	if result == nil {
		result = map[string]any{}
	}
	w.Header().Set("Content-Type", "application/json")
	if err := json.NewEncoder(w).Encode(result); err != nil {
		slog.Error("oauth store rpc encode failed", "op", req.Op, "error", err)
	}
}

// dispatch maps an RPC op to an OAuthStore method. Returns the JSON-encodable result (or nil → {}).
func (h *OAuthStoreHandler) dispatch(ctx context.Context, req *oauthRPCRequest) (any, error) {
	s := h.store
	switch req.Op {
	case "insertClient":
		return nil, s.InsertClient(ctx, req.Client.toEntity())
	case "getClientById":
		c, err := s.GetClientByID(ctx, req.ClientID)
		return map[string]any{"client": clientToDTO(c)}, err
	case "findClientsByName":
		clients, err := s.FindClientsByName(ctx, req.ClientName)
		dtos := make([]clientDTO, 0, len(clients))
		for i := range clients {
			dtos = append(dtos, *clientToDTO(&clients[i]))
		}
		return map[string]any{"clients": dtos}, err
	case "insertAuthCode":
		c := req.Code
		return nil, s.InsertAuthCode(ctx, &entity.OAuthAuthCode{
			CodeHash: c.CodeHash, ClientID: c.ClientID, SubjectKey: c.SubjectKey,
			RedirectURI: c.RedirectURI, CodeChallenge: c.CodeChallenge,
			CodeChallengeMethod: c.CodeChallengeMethod, Scope: c.Scope, ExpiresAt: c.ExpiresAt,
		})
	case "consumeAuthCode":
		code, err := s.ConsumeAuthCode(ctx, req.CodeHash)
		var dto *authCodeDTO
		if code != nil {
			dto = &authCodeDTO{
				CodeHash: code.CodeHash, ClientID: code.ClientID, SubjectKey: code.SubjectKey,
				RedirectURI: code.RedirectURI, CodeChallenge: code.CodeChallenge,
				CodeChallengeMethod: code.CodeChallengeMethod, Scope: code.Scope,
				ExpiresAt: code.ExpiresAt, ConsumedAt: code.ConsumedAt,
			}
		}
		return map[string]any{"code": dto}, err
	case "insertAccessToken":
		t := req.AccessToken
		return nil, s.InsertAccessToken(ctx, &entity.OAuthAccessToken{
			TokenHash: t.TokenHash, ClientID: t.ClientID, SubjectKey: t.SubjectKey,
			Scope: t.Scope, ExpiresAt: t.ExpiresAt,
		})
	case "getAccessTokenByHash":
		t, err := s.GetAccessTokenByHash(ctx, req.TokenHash)
		return map[string]any{"token": accessToDTO(t)}, err
	case "insertRefreshToken":
		t := req.RefreshToken
		id, err := s.InsertRefreshToken(ctx, &entity.OAuthRefreshToken{
			ID: t.ID, TokenHash: t.TokenHash, ClientID: t.ClientID, SubjectKey: t.SubjectKey,
			Scope: t.Scope, ExpiresAt: t.ExpiresAt,
		})
		return map[string]any{"id": id}, err
	case "getRefreshTokenByHash":
		t, err := s.GetRefreshTokenByHash(ctx, req.TokenHash)
		return map[string]any{"token": refreshToDTO(t)}, err
	case "consumeRefreshToken":
		consumed, err := s.ConsumeRefreshToken(ctx, req.ID)
		return map[string]any{"consumed": consumed}, err
	case "setRefreshRotatedTo":
		return nil, s.SetRefreshRotatedTo(ctx, req.ID, req.RotatedToID)
	case "revokeChain":
		return nil, s.RevokeChain(ctx, req.SubjectKey, req.ClientID)
	default:
		return nil, errUnknownOp
	}
}

var errUnknownOp = &opError{"unknown op"}

type opError struct{ msg string }

func (e *opError) Error() string { return e.msg }

File: apps/mcp/src/oauth-store.ts
import process from 'node:process';
import type {
  NewAccessToken,
  NewAuthCode,
  NewClient,
  NewRefreshToken,
  OAuthStore,
  StoredAccessToken,
  StoredAuthCode,
  StoredClient,
  StoredRefreshToken,
} from '@mcify/core';

/**
 * `OAuthStore` backed by the Lelemon Go backend. The MCP authorization server (mcify) owns the
 * OAuth protocol logic but persists nothing locally — every store call is an RPC to
 * `POST {LELEMON_ENDPOINT}/api/v1/internal/oauth`, authenticated with the shared `MCP_STORE_SECRET`.
 *
 * The wire shapes are the mcify `OAuthStore` row types verbatim (the Go DTOs use the same camelCase
 * field names), so this adapter is a thin pass-through — the only translation is dates: outbound
 * `Date`s serialize to ISO strings via `JSON.stringify`, and inbound ISO strings are revived to
 * `Date` on the fields the interface types as `Date`.
 */
export interface HttpOAuthStoreOptions {
  /** Backend base URL. Defaults to LELEMON_ENDPOINT or https://api.lelemon.dev. */
  endpoint?: string;
  /** Shared service secret (MCP_STORE_SECRET). */
  secret: string;
}

const DEFAULT_ENDPOINT = 'https://api.lelemon.dev';

type Json = Record<string, unknown>;

function reviveClient(c: Json | null): StoredClient | null {
  if (!c) return null;
  return { ...(c as unknown as StoredClient), createdAt: new Date(c['createdAt'] as string) };
}

function reviveAuthCode(c: Json | null): StoredAuthCode | null {
  if (!c) return null;
  return { ...(c as unknown as StoredAuthCode), expiresAt: new Date(c['expiresAt'] as string) };
}

function reviveAccess(t: Json | null): StoredAccessToken | null {
  if (!t) return null;
  const revoked = t['revokedAt'];
  return {
    ...(t as unknown as StoredAccessToken),
    expiresAt: new Date(t['expiresAt'] as string),
    revokedAt: revoked ? new Date(revoked as string) : null,
  };
}

function reviveRefresh(t: Json | null): StoredRefreshToken | null {
  if (!t) return null;
  const consumed = t['consumedAt'];
  return {
    ...(t as unknown as StoredRefreshToken),
    expiresAt: new Date(t['expiresAt'] as string),
    consumedAt: consumed ? new Date(consumed as string) : null,
  };
}

export class HttpOAuthStore implements OAuthStore {
  private readonly url: string;
  private readonly secret: string;

  constructor(options: HttpOAuthStoreOptions) {
    // The secret may be empty at construction (e.g. during `mcify build` with no env); it is
    // required only when a call is actually made — see `call()`.
    const base = options.endpoint ?? process.env['LELEMON_ENDPOINT'] ?? DEFAULT_ENDPOINT;
    this.url = `${base.replace(/\/+$/, '')}/api/v1/internal/oauth`;
    this.secret = options.secret;
  }

  private async call(op: string, payload: Json = {}): Promise<Json> {
    if (!this.secret) {
      throw new Error('MCP_STORE_SECRET is not configured — the MCP cannot reach the OAuth store.');
    }
    const res = await fetch(this.url, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${this.secret}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ op, ...payload }),
    });
    if (!res.ok) {
      const body = await res.text().catch(() => '');
      throw new Error(`OAuth store op "${op}" failed (${res.status}): ${body.slice(0, 200)}`);
    }
    return (await res.json()) as Json;
  }

  async insertClient(client: NewClient): Promise<StoredClient> {
    await this.call('insertClient', { client });
    // The backend returns {} on insert; re-read to get the canonical stored row (with createdAt).
    const stored = await this.getClientById(client.clientId);
    if (!stored) throw new Error('insertClient: client not found after insert');
    return stored;
  }

  async getClientById(clientId: string): Promise<StoredClient | null> {
    const out = await this.call('getClientById', { clientId });
    return reviveClient(out['client'] as Json | null);
  }

  async findClientsByName(clientName: string | null): Promise<StoredClient[]> {
    const out = await this.call('findClientsByName', { clientName: clientName ?? '' });
    const clients = (out['clients'] as Json[] | null) ?? [];
    return clients.map((c) => reviveClient(c)).filter((c): c is StoredClient => c !== null);
  }

  async insertAuthorizationCode(code: NewAuthCode): Promise<void> {
    await this.call('insertAuthCode', { code });
  }

  async consumeAuthorizationCode(codeHash: string): Promise<StoredAuthCode | null> {
    const out = await this.call('consumeAuthCode', { codeHash });
    return reviveAuthCode(out['code'] as Json | null);
  }

  async insertAccessToken(token: NewAccessToken): Promise<void> {
    await this.call('insertAccessToken', { accessToken: token });
  }

  async getAccessTokenByHash(tokenHash: string): Promise<StoredAccessToken | null> {
    const out = await this.call('getAccessTokenByHash', { tokenHash });
    return reviveAccess(out['token'] as Json | null);
  }

  async insertRefreshToken(token: NewRefreshToken): Promise<StoredRefreshToken> {
    const out = await this.call('insertRefreshToken', { refreshToken: token });
    return {
      id: out['id'] as string,
      clientId: token.clientId,
      subjectKey: token.subjectKey,
      scope: token.scope,
      expiresAt: token.expiresAt,
      consumedAt: null,
    };
  }

  async getRefreshTokenByHash(tokenHash: string): Promise<StoredRefreshToken | null> {
    const out = await this.call('getRefreshTokenByHash', { tokenHash });
    return reviveRefresh(out['token'] as Json | null);
  }

  async consumeRefreshToken(id: string): Promise<boolean> {
    const out = await this.call('consumeRefreshToken', { id });
    return out['consumed'] === true;
  }

  async setRefreshRotatedTo(id: string, rotatedToId: string): Promise<void> {
    await this.call('setRefreshRotatedTo', { id, rotatedToId });
  }

  async revokeChain(subjectKey: string, clientId: string): Promise<void> {
    await this.call('revokeChain', { subjectKey, clientId });
  }
}
